tests/unit/helpers/userCompanies.test.js

```diff
@@ -24,74 +24,212 @@ describe('create', () => {
     deleteManyCompanyLinkRequest.restore();
   });
 
-  it('should create UserCompany WITH startDate', async () => {
+  it('should create UserCompany, WITH startDate in payload', async () => {
     const userId = new ObjectId();
     const companyId = new ObjectId();
     const startDate = '2022-12-13T14:15:00.000Z';
+    const userCompanyStartDate = '2022-12-12T23:00:00.000Z';
 
     findOne.returns(SinonMongoose.stubChainedQueries(null, ['lean']));
 
     await UserCompaniesHelper.create({ user: userId, company: companyId, startDate });
 
-    sinon.assert.calledOnceWithExactly(create, { user: userId, company: companyId, startDate });
+    sinon.assert.calledOnceWithExactly(create, { user: userId, company: companyId, startDate: userCompanyStartDate });
     sinon.assert.calledOnceWithExactly(deleteManyCompanyLinkRequest, { user: userId });
     SinonMongoose.calledOnceWithExactly(
       findOne,
-      [{ query: 'findOne', args: [{ user: userId }, { company: 1 }] }, { query: 'lean' }]
+      [
+        {
+          query: 'findOne',
+          args: [
+            { user: userId, $or: [{ endDate: { $exists: false } }, { endDate: { $gt: userCompanyStartDate } }] },
+            { endDate: 1 },
+          ],
+        },
+        { query: 'lean' },
+      ]
     );
   });
 
-  it('should create UserCompany WITHOUT specified startDate', async () => {
+  it('should create UserCompany WITHOUT startDate in payload', async () => {
+    UtilsMock.mockCurrentDate('2020-12-10T10:11:12.134Z');
     const userId = new ObjectId();
     const companyId = new ObjectId();
 
     findOne.returns(SinonMongoose.stubChainedQueries(null, ['lean']));
 
     await UserCompaniesHelper.create({ user: userId, company: companyId });
 
-    sinon.assert.calledOnceWithExactly(create, { user: userId, company: companyId });
+    sinon.assert.calledOnceWithExactly(
+      create,
+      { user: userId, company: companyId, startDate: '2020-12-09T23:00:00.000Z' }
+    );
     sinon.assert.calledOnceWithExactly(deleteManyCompanyLinkRequest, { user: userId });
     SinonMongoose.calledOnceWithExactly(
       findOne,
-      [{ query: 'findOne', args: [{ user: userId }, { company: 1 }] }, { query: 'lean' }]
+      [
+        {
+          query: 'findOne',
+          args: [
+            { user: userId, $or: [{ endDate: { $exists: false } }, { endDate: { $gt: '2020-12-09T23:00:00.000Z' } }] },
+            { endDate: 1 },
+          ],
+        },
+        { query: 'lean' },
+      ]
     );
+    UtilsMock.unmockCurrentDate();
   });
 
-  it('should do nothing if already userCompany for this user and this company', async () => {
-    const userId = new ObjectId();
-    const companyId = new ObjectId();
+  it('should throw an error if already userCompany for this user and this company', async () => {
```

Comment:
je me pose la question sur ce cas fonctionnel :  est ce que ce cas et celui d'après ne sont pas inclus dans les 2 cas du dessous finalement ?
- Si le `userCompany` est en cours (pas de endDate) ou endDate apres la startDate proposée alors dans tous les cas on renvoie une erreur (que la structure soit la meme ou pas)
- Sinon, pourquoi ne pourrait on pas créer un nouveau `userCompany` pour cet apprenant ?

Reply:
et d'ailleurs je ne vois pas ou on fait la distinction dans le code du service

Reply:
J'ai hesite sur ces tests 🤔 Enfait avant on faisait la difference de si l'utilisateur avait la meme entreprise ou non : dans l'idee si quelqu'un essayait de rattacher un stagiaire a l'entreprise a laquelle il est deja rattache on s'etait dit que ca ne servait a rien d'afficher un message d'erreur. 

J'ai propose dans cette pr de le changer car maintenant comme on a une notion de temporalite, si on essayait d'ajouter pour un stagiaire dans l'entreprise A à partir du 20/12/2022, sachant qu'il etait deja dans la structure à partir du 23/12/2022, on rentrait dans un cas ou on ne changeait rien mais on n'affichait quand meme a l'utilisateur que le changement avait ete fait. 

Dans un second temps on pourra voir si ca ne modifie pas le userCompany plutot que de renvoyer un erreur mais ce ne me paraissait pas MVP vu qu'il fallait verifier plusieurs cas. Qu'en penses-tu ? 

Du coup, comme il y a un changement par rapport au comportement precedent, je me suis dit que c'etait bien de faire des tests dans les deux cas pour verifier que le changement etait bien effectif. (et pour documenter le fait qu'on a bien change ce comportement volontairement vu qu'on a ajoute des tests) Mais comme en meme temps on aurait pas fait de test si on etait parti directement sur cette logique je suis d'accord que ca pose question.

Qu'en penses-tu ? 
On peut s'appeler pour en parler :)

Reply:
on s'appelle oui :)

Reply:
Vu ensemble : on ne fait qu'une fois les tests en precisant dans le "it" que quelque soit l'entreprise (meme ou autre) on veut le meme resultat